### PR TITLE
Intrepid2: work on orientation for hierarchical basis and on new basis framewor

### PR DESCRIPTION
**Pull request by mperego**
@trilinos/intrepid2 

## Motivation

The current implementation of the Orientation did not preserve the structure of Hirarchical basis. This PR, among other things fixes this issue. List of changes

1. Added method `getSubCellBasis` to each basis, that return a pointer to a basis defined on a face or edge of the parent cell. (Thanks @CamelliaDPG  for useful suggestions). This allows to streamline the implementation of the Orientation Tools.

2. Work on Orientation to make it preserve the structure of hierarchical basis. The orientation is based on a reference basis defined on a subcells. If the basis defined on the subcell is "compatible" with the parent basis, than the Orientation mapping is essentially reordering and possibly scaling the basis functions, but it's not combining them.

3. Modify the "legacy" Lagrangian basis so that the orientation mapping is a signed permutation matrix. Note, this is not needed for the Orientiation to work, but it is a desirable property.

4. Added the EPointType (e.g. `EQUISPACED` or `WARPBLEND`)  as an argument of the derived basis, so that the derived basis can now fully include the Legacy Lagrangian basis

5. Added method for computing `DG` L2 projection, simply consisting in performing an L2 projection on each element (and possibly accounting for orientation).

## Testing

Several tests have been modified to exercise new capabilities.

### CODE REVIEW

**Comment by mperego** [comment]
@rppawlo  As you can see in commit 996bb7e, I had to modify a few regression values in a couple of Panzer tests. I didn't fully understand the tests but I think the differences are due to the fact that I changed the values of some low order basis and the related orientation mappings and those regression values are hard coded and assume a specific form of the basis. Let me know if this is fine.

Also, we might want to check Drekar and Empire before merging this.

**Comment by rppawlo** [comment]
@mperego - I don't think it is valid to change the field values in the panzer projection test. Can we get together over teams to talk about this later today? I will try to run the empire tests too.

**Comment by mperego** [comment]
> 
> 
> @mperego - I don't think it is valid to change the field values in the panzer projection test. Can we get together over teams to talk about this later today? I will try to run the empire tests too.

Thanks. Sure that would be great.

**Comment by rppawlo** [comment]
@mperego - just push a fix to this branch to remove compiler warnings - EMPIRE builds with warnings-as-error and these blocked the clang build. Once I got past this, just about every EMPIRE test is now failing at runtime. I think the same trick with scaling the edge length for the initial conditions for the edge basis is used all over the code.

**Comment by mperego** [comment]
> 
> 
> @mperego - just push a fix to this branch to remove compiler warnings - EMPIRE builds with warnings-as-error and these blocked the clang build. Once I got past this, just about every EMPIRE test is now failing at runtime. I think the same trick with scaling the edge length for the initial conditions for the edge basis is used all over the code.

@rppawlo Thanks! Let me fix Panzer projection as we discussed over then phone. Then we can decide whether to revert back 27c06a2 or to fix Empire. B.t.w. Are orientations in Empire hard-coded, or do they rely on the Orientation tools in Intrepid2?

**Comment by rppawlo** [comment]
I believe they are hard-coded, unless someone changed it recently.

**Comment by CamelliaDPG** [comment]
@mperego, sorry for the delay; I was away from email until just now.  I'll review this this morning.

**Comment by CamelliaDPG** [comment]
@mperego, a high-level question on these changes: do we make guarantees (or assumptions) about `getSubCellRefBasis()` being in the same order as it is in the original basis?  I.e., is there an implied mapping from the basis ordinals in the subcell basis to those in the whole-cell basis?

**Comment by mperego** [comment]
> 
> 
> @mperego, a high-level question on these changes: do we make guarantees (or assumptions) about `getSubCellRefBasis()` being in the same order as it is in the original basis? I.e., is there an implied mapping from the basis ordinals in the subcell basis to those in the whole-cell basis?

Yes, that was my intent

**Comment by mperego** [comment]
> 
> 
> > @mperego - just push a fix to this branch to remove compiler warnings - EMPIRE builds with warnings-as-error and these blocked the clang build. Once I got past this, just about every EMPIRE test is now failing at runtime. I think the same trick with scaling the edge length for the initial conditions for the edge basis is used all over the code.
> 
> @rppawlo Thanks! Let me fix Panzer projection as we discussed over then phone. Then we can decide whether to revert back [27c06a2](https://github.com/trilinos/Trilinos/commit/27c06a2c9d7edca5e56abbf013001d9756590898) or to fix Empire. B.t.w. Are orientations in Empire hard-coded, or do they rely on the Orientation tools in Intrepid2?

@rppawlo commit 89abfe4 fixes the issue w/ the projection in Panzer. Let me know if it makes sense to you. 

As for Empire, I think a quick fix could be to scale by a factor of 2 or 4 the coefficients of the HCurl and HDiv basis (I can list the exact scaling for every basis). Alternatively we can keep the low order basis as they are now, and modify only the higher order basis, which are not used outside of Panzer. I'd rather not do that, but it is a possibility. Anyway, before we take a decision let's wait until Nate is done with the review.

File: packages/intrepid2/unit-test/Projection/test_interpolation_projection_TET.hpp
Comment (CamelliaDPG) [inline packages/intrepid2/unit-test/Projection/test_interpolation_projection_TET.hpp:367]: Could switch this to a parallel_for (you'd need to make `vertices` a Kokkos::Array to get transparent lambda capture).

File: packages/panzer/adapters-stk/test/evaluator_tests/basis_times_vector.cpp
Comment (CamelliaDPG) [inline packages/panzer/adapters-stk/test/evaluator_tests/basis_times_vector.cpp:239]: Are the comments about the edge bases still correct, or are they off by a factor of 2 thanks to the change in basis definition?

File: packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefMatrixData.hpp
Comment (CamelliaDPG) [inline packages/intrepid2/src/Orientation/Intrepid2_OrientationToolsDefMatrixData.hpp:91]: This is much cleaner!  Thanks for making this change.

File: packages/intrepid2/src/Projection/Intrepid2_LagrangianInterpolationDef.hpp
Comment (CamelliaDPG) [inline packages/intrepid2/src/Projection/Intrepid2_LagrangianInterpolationDef.hpp:286]: Again, much cleaner.  Thank you!

**Review by CamelliaDPG — CHANGES_REQUESTED** [review]
@mperego, these changes look good to me overall.  You'll see that I've got some comments about making sure that the commit log properly captures the basis changes.  Feel free to defer the `parallel_for` additions I suggested in the tests; perhaps those should wait until we're focusing on the UVM cleanup.  But please do check the comments in the panzer tests where I indicated to see whether they need to be updated to reflect your changes.

I'm happy to Approve whenever you're ready, but I'm marking this review as "Request changes" to make sure you have a chance to make changes before merge if you want to.

**Comment by mperego** [comment]
@CamelliaDPG Thanks a  lot for your review, I'll work on that tomorrow. 
In the meantime, I'd like to ask you what you think about the fact that I changed the basis definitions (commit 27c06a2c). Here is the motivation (commit message):
```
Intrepid2: Modify Reference bases so that the orientation matrix is a signed permuatation

Although not required for the proper working of the orientation tools,
it is desirable to have a orientation matrix that is simply a signed permutation.
In order for this to be true, we need that the reference bases defined on subcells are
equivalent to the restriction (projected on the tangent or normal manifold to the subcell for HCURL and HDIV respectively)
of the parent cell basis.

Some of the legacy Intrepid2 Lagrangian bases did not comply with this and have been fixed.

Also renaming the EBase enum  BASIS_FEM_FIAT to BASIS_FEM_LAGRANGIAN, as the modified bases
are no longer equivalent to  FIAT bases
```

So, the reason why I redefined the basis is so that the Orientation mapping is a signed permutation, without requiring specialized treating of HCURL and HDIV low-order bases and HCURL and HDIV Tet and Triangle bases.
I think this makes the code cleaner and more uniform.
However, Empire is apparently making assumption on the low-order bases definition and we would need to fix that. I'd be in favor of that but I'd like to hear your input on this.

**Comment by CamelliaDPG** [comment]
Thanks for explaining the motivation; that makes sense. 

> I think this makes the code cleaner and more uniform.

Yes, I agree, and this is demonstrated too by the number of places in your PR where you were able to delete code that weighted edge values by a factor of 2 or whatever.

> However, Empire is apparently making assumption on the low-order bases definition and we would need to fix that. I'd be in favor of that but I'd like to hear your input on this.

Yes, I think it's well worth it to make those fixes in EMPIRE.

**Comment by mperego** [comment]
@rppawlo, Nate and I think it is worth fixing Empire to use the new definition of the low-order bases. What do you think? If you give me access to Empire and point me to scripts to build it, I can have a look at it and see if I can fix that.

**Comment by mperego** [comment]
> 
> 
> @rppawlo, Nate and I think it is worth fixing Empire to use the new definition of the low-order bases. What do you think? If you give me access to Empire and point me to scripts to build it, I can have a look at it and see if I can fix that.

@rppawlo Pinging you on this. The changes requested by Nate are minor and I'm waiting to hear whether you are fine with the changes to the basis so that I can do those changes and merge the branch. As I mentioned, I'd be happy to work in Empire and see if I can fix it quickly.

**Comment by rppawlo** [comment]
@mperego - I'd prefer that we not merge this in until we have a fix on a branch that is ready for empire to merge. I'm worried that this is not going to be a simple few places that we easily find. Additionally we have not tested Drekar but I am sure we will find same issues there. This is a significant break in backwards compatibility that could be disruptive.

**Comment by mperego** [comment]
> 
> 
> @mperego - I'd prefer that we not merge this in until we have a fix on a branch that is ready for empire to merge. I'm worried that this is not going to be a simple few places that we easily find. Additionally we have not tested Drekar but I am sure we will find same issues there. This is a significant break in backwards compatibility that could be disruptive.

@rppawlo I was not proposing to merge it before having a fix for Empire. As I mentioned, I'd be happy to work on a fix. Would you like me to work on that?

**Comment by mperego** [comment]
> 
> 
> > @mperego - I'd prefer that we not merge this in until we have a fix on a branch that is ready for empire to merge. I'm worried that this is not going to be a simple few places that we easily find. Additionally we have not tested Drekar but I am sure we will find same issues there. This is a significant break in backwards compatibility that could be disruptive.
> 
> @rppawlo I was not proposing to merge it before having a fix for Empire. As I mentioned, I'd be happy to work on a fix. Would you like me to work on that?

As for Drekar, do you think somebody from the Drekar team, e.g. @sconde,  can look into that? I can then work with them and find a fix.

File: packages/intrepid2/unit-test/Projection/test_interpolation_projection_QUAD.hpp
Comment (mperego) [reply packages/intrepid2/unit-test/Projection/test_interpolation_projection_QUAD.hpp:1536]: @CamelliaDPG That is a nice suggestion. At the moment this test (as long as other you pointed out) has been implemented under the assumption that UVM is enable. There are several parts of the test that would need to be changed and here added a check for L2 projection, following the structure of the test. 
I think it makes more sense to address this when we work on making Intrepid2 UVM free. But I'll keep your comment in mind for the future.

File: packages/panzer/adapters-stk/test/evaluator_tests/basis_times_vector.cpp
Comment (mperego) [reply packages/panzer/adapters-stk/test/evaluator_tests/basis_times_vector.cpp:239]: Good catch. I'll fix the comments.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_HCURL_HEX_I1_FEMDef.hpp
Comment (mperego) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_HCURL_HEX_I1_FEMDef.hpp:74]: The commit message indicate that the legacy basis function changed. I can expand it explicitly listing all the bases that changed

Comment (CamelliaDPG) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_HCURL_HEX_I1_FEMDef.hpp:74]: Yes, I think that would be good.  Thanks!

File: packages/intrepid2/unit-test/Projection/test_interpolation_projection_QUAD.hpp
Comment (CamelliaDPG) [reply packages/intrepid2/unit-test/Projection/test_interpolation_projection_QUAD.hpp:1536]: @mperego Sounds good to me!

**Comment by rppawlo** [comment]
@mperego - sure. If you want to look into the empire issues, you'll have to contact @jmgate to get started with the code. I'll open a ticket an flag the right people. If @sconde  can't make the changes to Drekar, I can take a look.

**Comment by sconde** [comment]
Looking at this now

**Comment by mperego** [comment]
> 
> 
> @mperego - sure. If you want to look into the empire issues, you'll have to contact @jmgate to get started with the code. I'll open a ticket an flag the right people. If @sconde can't make the changes to Drekar, I can take a look.

Thanks @rppawlo ! @jmgate  just gave me access to Empire. I hope I'm not shooting myself int he foot ;-)

**Comment by sconde** [comment]
Drekar's exodiff tests are failing with this PR:

```
980 The following tests FAILED:                                                                                                                                                                                                                                                                                                                                           
981     396 - Drekar_divergence_error_LEP (Failed)                                                                                                                                                                                                                                                                                                                        
982     431 - Drekar_plasma_oscillation (Failed)                                                                                                                                                                                                                                                                                                                          
983     432 - Drekar_plasma_oscillation_rtc (Failed)                                                                                                                                                                                                                                                                                                                      
984     433 - Drekar_plasma_oscillation_rtc_2D (Failed)                                                                                                                                                                                                                                                                                                                   
985     434 - Drekar_plasma_oscillation_ES (Failed)                                                                                                                                                                                                                                                                                                                       
986     435 - Drekar_plasma_oscillation_ES_2D (Failed)                                                                                                                                                                                                                                                                                                                    
987     436 - Drekar_plasma_oscillation_tetmesh (Failed)                                                                                                                                                                                                                                                                                                                  
988     437 - Drekar_braginskii_zeroB (Failed)                                                                                                                                                                                                                                                                                                                            
989     438 - Drekar_braginskii_strongB (Failed)                                                                                                                                                                                                                                                                                                                          
990     439 - Drekar_current_drive (Failed)                                                                                                                                                                                                                                                                                                                               
991     440 - Drekar_standing_wave (Failed)                                                                                                                                                                                                                                                                                                                               
992     441 - Drekar_impedance_bc (Failed)                                                                                                                                                                                                                                                                                                                                
993     442 - Drekar_elc_standing_wave (Failed)                                                                                                                                                                                                                                                                                                                           
994     443 - Drekar_mag_standing_wave (Failed)                                                                                                                                                                                                                                                                                                                           
995     445 - Drekar_maxwell_potential_incoming_wave (Failed) 
```

Inspecting the plasma oscillation test, we see the difference is significant. 

![density-gold-vs-pr](https://user-images.githubusercontent.com/1556219/94580434-eb630200-0247-11eb-903c-91c881403956.png)

**Comment by mperego** [comment]
> 
> 
> Drekar's exodiff tests are failing with this PR:
> 
> ```
> 980 The following tests FAILED:                                                                                                                                                                                                                                                                                                                                           
> 981     396 - Drekar_divergence_error_LEP (Failed)                                                                                                                                                                                                                                                                                                                        
> 982     431 - Drekar_plasma_oscillation (Failed)                                                                                                                                                                                                                                                                                                                          
> 983     432 - Drekar_plasma_oscillation_rtc (Failed)                                                                                                                                                                                                                                                                                                                      
> 984     433 - Drekar_plasma_oscillation_rtc_2D (Failed)                                                                                                                                                                                                                                                                                                                   
> 985     434 - Drekar_plasma_oscillation_ES (Failed)                                                                                                                                                                                                                                                                                                                       
> 986     435 - Drekar_plasma_oscillation_ES_2D (Failed)                                                                                                                                                                                                                                                                                                                    
> 987     436 - Drekar_plasma_oscillation_tetmesh (Failed)                                                                                                                                                                                                                                                                                                                  
> 988     437 - Drekar_braginskii_zeroB (Failed)                                                                                                                                                                                                                                                                                                                            
> 989     438 - Drekar_braginskii_strongB (Failed)                                                                                                                                                                                                                                                                                                                          
> 990     439 - Drekar_current_drive (Failed)                                                                                                                                                                                                                                                                                                                               
> 991     440 - Drekar_standing_wave (Failed)                                                                                                                                                                                                                                                                                                                               
> 992     441 - Drekar_impedance_bc (Failed)                                                                                                                                                                                                                                                                                                                                
> 993     442 - Drekar_elc_standing_wave (Failed)                                                                                                                                                                                                                                                                                                                           
> 994     443 - Drekar_mag_standing_wave (Failed)                                                                                                                                                                                                                                                                                                                           
> 995     445 - Drekar_maxwell_potential_incoming_wave (Failed) 
> ```
> 
> Inspecting the plasma oscillation test, we see the difference is significant.
> 
> ![density-gold-vs-pr](https://user-images.githubusercontent.com/1556219/94580434-eb630200-0247-11eb-903c-91c881403956.png)

Thanks @sconde . That's what @rppawlo predicted. 
It seems that Drekar is assuming a specific scaling for the bases and we changed that in this PR. 
That's not the correct way to use Intrepid2 but it was the easiest way to go, especially when Intrepid2 didn't provide projections.

What HCURL and HDIV basis functions are you using in EMPIRE? Only the low order ones?

**Comment by mperego** [comment]
@sconde I made a few changes to Panzer. When you have time can you retry testing Drekar?

**Comment by mperego** [comment]
@rppawlo I realized I had to fix `ProjectToFaces` and `ProjectToEdges`, see 6bb1762c.
At the moment they only work for low order basis, but it should not take long to use projections and make them work for every order. 

(I think I fixed most of the Empire tests, more tomorrow)

**Comment by mperego** [comment]
@CamelliaDPG I think I did all the changes we discussed. I also modified the low order HCurl and HDiv Wedge elements as I forgot to do so before. I also modified a few more classes in Panzer (see last commits) that are used by Empire. I rebased the commits.

**Comment by CamelliaDPG** [comment]
> @CamelliaDPG I think I did all the changes we discussed. I also modified the low order HCurl and HDiv Wedge elements as I forgot to do so before. I also modified a few more classes in Panzer (see last commits) that are used by Empire. I rebased the commits.

@mperego, sounds good.  Have you (or has someone) confirmed that EMPIRE tests pass now with these changes?

**Comment by mperego** [comment]
> 
> 
> > @CamelliaDPG I think I did all the changes we discussed. I also modified the low order HCurl and HDiv Wedge elements as I forgot to do so before. I also modified a few more classes in Panzer (see last commits) that are used by Empire. I rebased the commits.
> 
> @mperego, sounds good. Have you (or has someone) confirmed that EMPIRE tests pass now with these changes?

I finishing working on a PR for Empire. I think I fixed all the tests I was able to run (they were 2 tests that were failing before, probably because of my environment). Anyway, Matt offered to check it when we are done.

File: packages/intrepid2/src/Discretization/Basis/Intrepid2_HCURL_TET_In_FEM.hpp.orig
Comment (CamelliaDPG) [inline packages/intrepid2/src/Discretization/Basis/Intrepid2_HCURL_TET_In_FEM.hpp.orig:1]: @mperego, I think you might have included this file by accident.  (It has some git merge conflict markings in it.)

Comment (mperego) [reply packages/intrepid2/src/Discretization/Basis/Intrepid2_HCURL_TET_In_FEM.hpp.orig:1]: Argh! Good catch

File: packages/panzer/disc-fe/src/evaluators/Panzer_ProjectToFaces_impl.hpp
Comment (CamelliaDPG) [inline packages/panzer/disc-fe/src/evaluators/Panzer_ProjectToFaces_impl.hpp:124]: @mperego, it looks like the earlier code here supported higher-order bases.  Do we need to make the new version support higher-order as well?

Comment (mperego) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ProjectToFaces_impl.hpp:124]: @CamelliaDPG no, it was not supporting high order basis. I think it was only an attempt in that direction, but we would do it differently now, using projection. As a matter of fact the code assumes to have that only one evaluation point per face, which make sense only for constants, i.e. for low order basis.

Comment (CamelliaDPG) [reply packages/panzer/disc-fe/src/evaluators/Panzer_ProjectToFaces_impl.hpp:124]: OK, sounds good!

**Review by CamelliaDPG — APPROVED** [review]
@mperego, looks good to me.  Please do delete that extra file before merge.  Thanks!

**Comment by mperego** [comment]
@rppawlo Are you OK with the changes to Panzer? (commit 8fde418)

@bathmatt is fine if we merge this. What about Drekar? There is still one test failing there as well. Should we wait for it to be fixed?

**Pull request merged**
